### ISSUE TITLE
Migrating to zod 3 - ZodType(?) issue

### ISSUE THREAD

**Issue by azhiv** [colinhacks/zod#628]
We've been using zod 2 with great success, but now I'm having an issue migrating to zod 3. Here's the piece of code representing the error:
```ts
import * as zod from 'zod';

const schema = zod.object({
  videoDate: zod.string(),
});

class Editor<
  S extends zod.ZodType<zod.TypeOf<typeof schema>> = typeof schema,
> {
  constructor(public schema: S) {}
}

const advancedSchema = zod.intersection(schema,
  zod.object({
    videoDuration: zod.number(),
  })
    .refine(val => val.videoDuration > 5),
);

export class AdvancedEditor extends Editor<typeof advancedSchema> {      // the error pops here
  constructor(schema: typeof advancedSchema) {
    super(schema);
  }
}
```

This was legitimate in [zod 2 ](https://codesandbox.io/s/gifted-lumiere-ur9k9?file=/src/index.ts), but turned out to be prohibited in [zod 3](https://codesandbox.io/s/eager-cdn-7kute?file=/src/index.ts) yielding the following error:
```
Type 'ZodIntersection<ZodObject<{ videoDate: ZodString; }, "strip", ZodTypeAny, { videoDate?: string; }, { videoDate?: string; }>, ZodEffects<ZodObject<{ videoDuration: ZodNumber; }, "strip", ZodTypeAny, { ...; }, { ...; }>, { ...; }>>' does not satisfy the constraint 'ZodType<{ videoDate?: string; }, ZodTypeDef, { videoDate?: string; }>'.
  The types returned by '_parse(...)' are incompatible between these types.
    Type 'ParseReturnType<ZodObject<{ videoDate: ZodString; }, "strip", ZodTypeAny, { videoDate?: string; }, { videoDate?: string; }> & ZodEffects<ZodObject<{ videoDuration: ZodNumber; }, "strip", ZodTypeAny, { ...; }, { ...; }>, { ...; }>>' is not assignable to type 'ParseReturnType<{ videoDate?: string; }>'.
      Type 'OK<ZodObject<{ videoDate: ZodString; }, "strip", ZodTypeAny, { videoDate?: string; }, { videoDate?: string; }> & ZodEffects<ZodObject<{ videoDuration: ZodNumber; }, "strip", ZodTypeAny, { ...; }, { ...; }>, { ...; }>>' is not assignable to type 'ParseReturnType<{ videoDate?: string; }>'.
        Type 'OK<ZodObject<{ videoDate: ZodString; }, "strip", ZodTypeAny, { videoDate?: string; }, { videoDate?: string; }> & ZodEffects<ZodObject<{ videoDuration: ZodNumber; }, "strip", ZodTypeAny, { ...; }, { ...; }>, { ...; }>>' is not assignable to type 'OK<{ videoDate?: string; }>'.
          Type 'ZodObject<{ videoDate: ZodString; }, "strip", ZodTypeAny, { videoDate?: string; }, { videoDate?: string; }> & ZodEffects<ZodObject<{ videoDuration: ZodNumber; }, "strip", ZodTypeAny, { ...; }, { ...; }>, { ...; }>' has no properties in common with type '{ videoDate?: string; }'.ts(2344)
```
Can I follow the same approach now?

**Comment by scotttrinh** [comment]
`ZodIntersection` does not extend `ZodType` like that. For this specific case if you can use `extend` instead of `intersection`, it should work:

```ts
import * as zod from "zod";

const schema = zod.object({
  videoDate: zod.string()
});

class Editor<S extends zod.ZodType<zod.output<typeof schema>>> {
  constructor(public schema: S) {}
}

const advancedSchema = schema.extend({
  videoDuration: zod.number().min(5)
});

export class AdvancedEditor extends Editor<typeof advancedSchema> {
  constructor(schema: typeof advancedSchema) {
    super(schema);
  }
}
```

**Comment by azhiv** [comment]
But you've lost the `refine` which is vital. I altered the example a little bit to demonstrate the obstacle:
```ts
const advancedSchema = zod.intersection(
  schema,
  zod
    .object({
      videoDuration: zod.number(),
      valueToCompare: zod.number(),
    })
    .refine((val) => val.videoDuration > val.valueToCompare)
);
```

Just a side note, in zod 2 I was also able to define `advancedSchema` as follows:
```ts
const advancedSchema = schema.merge(
  zod
    .object({
      videoDuration: zod.number(),
      valueToCompare: zod.number(),
    })
    .refine((val) => val.videoDuration > val.valueToCompare)
);
```
But this is no longer allowed in zod 3.

**Comment by scotttrinh** [comment]
Oh, sorry, I only dropped the refine because it seemed unnecessary given the simplicity of it. You can refine an `extend`ed schema, too:

```ts
import * as zod from "zod";

const schema = zod.object({
  videoDate: zod.string()
});

class Editor<S extends zod.ZodType<zod.output<typeof schema>>> {
  constructor(public schema: S) {}
}

const advancedSchema = schema
  .extend({
    videoDuration: zod.number(),
    valueToCompare: zod.number()
  })
  .refine((val) => val.videoDuration > val.valueToCompare);

export class AdvancedEditor extends Editor<typeof advancedSchema> {
  constructor(schema: typeof advancedSchema) {
    super(schema);
  }
}
```

**Comment by colinhacks** [comment]
@azhiv  This actually was a bug... Fixed in `zod@3.8.2`

The return type of ZodIntersection._parse was wrong. This must have been changed in the parser rewrite and no one caught it. Intersections aren't very commonly used, and in general I'd recommending moving refinement to the top-level if possible (though this is a stylistic preference and it should work fine your way).

**Comment by azhiv** [comment]
Thank you!
